contracts/c-sphincs-all-in-one-lock/utils/ckb_tx_message_all.h

```diff
@@ -0,0 +1,225 @@
+#ifndef CKB_CKB_TX_MESSAGE_ALL_H_
+#define CKB_CKB_TX_MESSAGE_ALL_H_
+
+#include <ckb_syscalls.h>
+
+#include "witness_args_lazy_utils.h"
+
+#ifndef CKB_TX_MESSAGE_WITNESS_BUFFER_SIZE
+#define CKB_TX_MESSAGE_WITNESS_BUFFER_SIZE (1024 * 64)
+#endif
+
+#ifndef CKB_TX_MESSAGE_CKB_LOAD_BUFFER_SIZE
+#define CKB_TX_MESSAGE_CKB_LOAD_BUFFER_SIZE (1024 * 64)
+#endif
+
+typedef int (*ckb_tx_message_write_func_t)(const uint8_t* data, size_t length,
+                                           void* context);
+typedef int (*_ckb_tx_message_ckb_load_func_t)(void* addr, uint64_t* len,
+                                               size_t offset, size_t index,
+                                               size_t source);
+
+static int _ckb_tx_message_load_and_hash(_ckb_tx_message_ckb_load_func_t loader,
+                                         int sign_length, size_t index,
+                                         size_t source,
+                                         ckb_tx_message_write_func_t writer,
+                                         void* context);
+
+static int ckb_tx_message_all_generate_with_witness_args(
+    ckb_tx_message_write_func_t writer, void* context,
+    WitnessArgsType* first_witness) {
+  int err;
+
+  /* Hash tx hash */
+  {
+    uint8_t tx_hash[32];
+    uint64_t len = 32;
+    err = ckb_load_tx_hash(tx_hash, &len, 0);
+    if (err != 0) {
+      return err;
+    }
+    ASSERT(len == 32);
+    err = writer(tx_hash, 32, context);
+    if (err != 0) {
+      return err;
+    }
+  }
+
+  /* Hash all input cells */
+  size_t input_cells = 0;
+  {
+    while (1) {
+      err = _ckb_tx_message_load_and_hash(ckb_load_cell, 0, input_cells,
+                                          CKB_SOURCE_INPUT, writer, context);
+      if (err == CKB_INDEX_OUT_OF_BOUND) {
+        break;
+      }
+      if (err != 0) {
+        return err;
+      }
+
+      err = _ckb_tx_message_load_and_hash(ckb_load_cell_data, 1, input_cells,
+                                          CKB_SOURCE_INPUT, writer, context);
+      if (err != 0) {
+        return err;
+      }
+      input_cells += 1;
+    }
+  }
+
+  /* Hash the first witness in specified format */
+  {
+    uint8_t buffer[CKB_TX_MESSAGE_CKB_LOAD_BUFFER_SIZE];
+    {
+      uint32_t read_len = mol2_read_at(&first_witness->cur, buffer, 16);
+      if (read_len != 16) {
+        MOL2_PANIC(MOL2_ERR_DATA);
+        return MOL2_ERR_DATA;
+      }
+      err = writer(buffer, 16, context);
+      if (err != 0) {
+        return err;
+      }
+    }
+
+    {
```

Comment:
if `_ckb_tx_message_load_and_hash` can accept `offset` argument, the following 2 code blocks can use this function. Note, the type mol2_cursor_t has an `offset` field.

Reply:
I'm not sure how I feel about it: yes the following 2 blocks can be minimized to a single `_ckb_tx_message_load_and_hash`, but the code will rely on a hidden knowledge, that `input_type` and `output_type` are next to each other, and there are no data afterwards, this is true but still a hidden knowledge, which is also not what the `CKB_TX_MESSAGE_ALL` spec defines. I wonder if people would be confused by the change.

If there will be major gains in terms of performance or code size, I would like to make the change, otherwise, I personally prefer the currenct code since it explicitly tells what are being hashed.